athenahealth/lab_results.go

```diff
@@ -0,0 +1,303 @@
+package athenahealth
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/url"
+	"strconv"
+	"time"
+)
+
+type LabResult struct {
+	AttachmentExists         bool   `json:"attachmentexists"`
+	Description              string `json:"description"`
+	ExactDuplicateDocumentID int    `json:"exactduplicatedocumentid"`
+	FacilityID               int    `json:"facilityid"`
+	InternalNote             string `json:"internalnote"`
+	IsReviewedByProvider     string `json:"isreviewedbyprovider"`
+	LabResultDate            string `json:"labresultdate"`
+	LabResultDateTime        string `json:"labresultdatetime"`
+	LabResultID              int    `json:"labresultid"`
+	LabResultLOINC           string `json:"labresultloinc"`
+	LabResultNote            string `json:"labresultnote"`
+	OrderID                  int    `json:"orderid"`
+	PatientNote              string `json:"patientnote"`
+	Priority                 string `json:"priority"`
+	ProviderID               int    `json:"providerid"`
+	ResultStatus             string `json:"resultstatus"`
+}
+
+type ListLabResultsOptions struct {
+	StartDate           *time.Time `json:"startdate"`
+	LabResultStatus     *string    `json:"labresultstatus"`
+	ShowHidden          *bool      `json:"showhidden"`
+	ShowAbnormalDetails *bool      `json:"showabnormaldetails"`
+	EndDate             *time.Time `json:"enddate"`
+	HideDuplicate       *bool      `json:"hideduplicate"`
+
+	Pagination *PaginationOptions
+}
+
+type listLabResultsResponse struct {
+	LabResults []*LabResult `json:"results"`
+
+	PaginationResponse
+}
+
+type ListLabResultsResult struct {
+	LabResults []*LabResult
+
+	Pagination *PaginationResult
+}
+
+// ListLabResults retrieves the patients laboratory results
+//
+// GET /v1/{practiceid}/chart/{patientid}/labresults
+//
+// https://docs.athenahealth.com/api/api-ref/lab-result#Get-patient's-lab-results
+func (h *HTTPClient) ListLabResults(ctx context.Context, patientID string, departmentID string, opts *ListLabResultsOptions) (*ListLabResultsResult, error) {
```

Comment:
assuming that we're all on board with the proposal in this other PR https://github.com/eleanorhealth/go-athenahealth/pull/111, we should keep non-url path parameters that are required in the options struct as non-pointer fields.

Reply:
Ah interesting, I missed that. Does make sense, and we should probably error if not set 👍

Reply:
We discussed live with the team and agreed we'd like to use a more enforced version of the current intended pattern of passing all required params as method arguments and optional params as an optional (pointer to) options struct, plus enforcing required field presence where it makes sense before making the call to athena. I've updated my branch to reflect this and @dillonstreator will update external parties / the README to reflect the decision and pattern.

Reply:
https://github.com/eleanorhealth/go-athenahealth#method-signatures-required-vs-optional-fields